src/zarr/core/indexing.py

```diff
@@ -34,7 +34,9 @@
 IntSequence = list[int] | npt.NDArray[np.intp]
 ArrayOfIntOrBool = npt.NDArray[np.intp] | npt.NDArray[np.bool_]
 BasicSelector = int | slice | EllipsisType
-Selector = BasicSelector | ArrayOfIntOrBool
+Selector = (
+    BasicSelector | ArrayOfIntOrBool | Array
+)  # help! we want zarr.Array[np.intp] | zarr.Array[np.bool_]
```

Comment:
@dstansby and @normanrz - I believe we want something like I've written here but am not clear if we can do this yet. Any ideas?

Reply:
I don't see a technical reason why it isn't possible to make `Array` a generic type over data types.

Reply:
happy to do this, can you point me in the right direction?

Reply:
@dstansby - I feel that you are going to be best positioned to do flesh out the generic Array type. Do you have time to take that on?

Reply:
Probably not (at least in the next month) :(